### PR TITLE
Metrics javadoc fixes #767

### LINKED ISSUE #767: Clarify / Document metrics contract
The metrics contract is a bit unclear, from the implementation. Since it's not defined in the spec, having the only fully implemented metrics for Parquet, and while I'm working on ORC metrics it's not very clear what is the contract expected since file formats seem to implement this differently, for instance:

* `Map<Integer, Long> valueCounts()` - it's not clear whether this method includes non-null or repeated values. As per the `TestMetrics` it looks like value counts *includes null and repeated values* which would be pretty much the same as row count, except for nested structures (e.g. lists, maps) - however this is not defined.

This issue is to track the discussion about the expected metrics contract and get a clear definition.

### PR DESCRIPTION
**Pull request by edgarRd**
Adding javadoc to `Metrics` following `Parquet` metrics implementation as proposal.

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/TestMetrics.java [hunk 1]
```diff
@@ -114,6 +114,56 @@ public abstract File writeRecords(Schema schema, Map<String, String> properties,
...
+    secondRecord.put("fixedCol", fixed);
+    secondRecord.put("binaryCol", "S".getBytes());
+
+    File recordsFile = writeRecords(SIMPLE_SCHEMA, firstRecord, secondRecord);
+
+    Metrics metrics = getMetrics(Files.localInput(recordsFile));
+    Assert.assertEquals(2L, (long) metrics.recordCount());
+    assertCounts(1, 2L, 0L, metrics);
+    assertCounts(2, 2L, 0L, metrics);
+    assertCounts(3, 2L, 2L, metrics);
```
Comment (edgarRd) [inline core/src/test/java/org/apache/iceberg/TestMetrics.java]: Note that in this case `valueCounts == nullValueCounts` and the count is the same as `recordCount`.

**Comment by edgarRd** [comment]
PTAL @rdsr @rdblue  - Thanks!

File: api/src/main/java/org/apache/iceberg/Metrics.java [hunk 2]
```diff
@@ -59,26 +62,56 @@ public Metrics(Long rowCount,
     this.upperBounds = upperBounds;
   }
 
+  /**
+   * Get the number of records (rows) in file.
+   *
+   * @return
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/Metrics.java]: Nit: should not be left blank or else we get javadoc warnings.

File: api/src/main/java/org/apache/iceberg/Metrics.java [hunk 3]
```diff
@@ -59,26 +62,56 @@ public Metrics(Long rowCount,
...
+   * @return a Map of fieldId to the size in bytes
+   */
   public Map<Integer, Long> columnSizes() {
     return columnSizes;
   }
 
+  /**
+   * Get the number of all values, including null and repeated.
+   *
+   * @return a Map of fieldId to the number of all values
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/Metrics.java]: All values including nulls and NaN?

File: api/src/main/java/org/apache/iceberg/Metrics.java [hunk 4]
```diff
@@ -59,26 +62,56 @@ public Metrics(Long rowCount,
...
+   * @return a Map of fieldId to the number of nulls
+   */
   public Map<Integer, Long> nullValueCounts() {
     return nullValueCounts;
   }
 
+  /**
+   * Get the non-null lower bound values for all fields in a file.
+   *
+   * @return a Map of fieldId to the lower bound value as a ByteBuffer
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/Metrics.java]: Can we also add a reference to the conversion back to a value? And maybe one to the spec for that conversion?

File: core/src/test/java/org/apache/iceberg/TestMetrics.java [hunk 5]
```diff
@@ -114,6 +114,56 @@ public abstract File writeRecords(Schema schema, Map<String, String> properties,
 
   public abstract int splitCount(File parquetFile) throws IOException;
 
+  @Test
+  public void testMetricsForRepeatedValues() throws IOException {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestMetrics.java]: What do you want to test with this case?

**Comment by rdblue** [comment]
Mostly looks good, but I don't understand the motivation for the new test case.

File: api/src/main/java/org/apache/iceberg/Metrics.java (hunk 4 above)
Comment (edgarRd) [reply api/src/main/java/org/apache/iceberg/Metrics.java]: Done.

File: core/src/test/java/org/apache/iceberg/TestMetrics.java (hunk 5 above)
Comment (edgarRd) [reply core/src/test/java/org/apache/iceberg/TestMetrics.java]: This test validates the contract that counts include repeated values. In my ORC implementation counts were not included and there was no test validating that contract that now I've documented.

**Comment by edgarRd** [comment]
cc @aokolnychyi

**Comment by rdblue** [comment]
Thanks for clarifying what was being tested! I've merged this.

**Pull request merged**